File: neptun_alarm_app/lib/services/ballistic_alert_service.dart
import 'package:flutter/material.dart';

/// Сервіс для керування балістичними тривогами
/// Дозволяє показувати ефектні оповіщення на карті
class BallisticAlertService {
  static final BallisticAlertService _instance = BallisticAlertService._internal();
  factory BallisticAlertService() => _instance;
  BallisticAlertService._internal();

  // Callbacks для оповіщення UI
  final List<void Function(String? region)> _threatCallbacks = [];
  final List<void Function(String? region)> _allClearCallbacks = [];
  
  // Поточний стан
  bool _isBallisticThreatActive = false;
  String? _currentRegion;
  DateTime? _lastThreatTime;
  
  bool get isBallisticThreatActive => _isBallisticThreatActive;
  String? get currentRegion => _currentRegion;
  DateTime? get lastThreatTime => _lastThreatTime;
  
  /// Підписатися на оповіщення про балістичну загрозу
  void onBallisticThreat(void Function(String? region) callback) {
    _threatCallbacks.add(callback);
  }
  
  /// Підписатися на оповіщення про відбій
  void onBallisticAllClear(void Function(String? region) callback) {
    _allClearCallbacks.add(callback);
  }
  
  /// Відписатися від оповіщень
  void removeCallback(void Function(String? region) callback) {
    _threatCallbacks.remove(callback);
    _allClearCallbacks.remove(callback);
  }
  
  /// Мінімальний час (в секундах) перед тим як відбій може перекрити загрозу
  static const int _minThreatDurationSeconds = 10;
  
  /// Викликати балістичну загрозу
  void triggerBallisticThreat({String? region}) {
    _isBallisticThreatActive = true;
    _currentRegion = region;
    _lastThreatTime = DateTime.now();
    
    debugPrint('🚀 BallisticAlertService: Triggering threat for $region at $_lastThreatTime');
    
    for (final callback in _threatCallbacks) {
      callback(region);
    }
  }
  
  /// Викликати відбій балістичної загрози
  void triggerBallisticAllClear({String? region}) {
    // Захист: якщо загроза була менше ніж _minThreatDurationSeconds секунд тому,
    // ігноруємо відбій (щоб не перекривати загрозу)
    if (_lastThreatTime != null) {
      final elapsed = DateTime.now().difference(_lastThreatTime!).inSeconds;
      if (elapsed < _minThreatDurationSeconds) {
        debugPrint('⚠️ BallisticAlertService: Ignoring all-clear, threat was only ${elapsed}s ago (min: ${_minThreatDurationSeconds}s)');
        return;
      }
    }
    
    _isBallisticThreatActive = false;
    _currentRegion = null;
    
    debugPrint('✅ BallisticAlertService: All clear for $region');
    
    for (final callback in _allClearCallbacks) {
      callback(region);
    }
  }
  
  /// Перевірити чи це балістичне оповіщення за текстом
  static BallisticAlertType? detectAlertType(String message) {
    final lowerMessage = message.toLowerCase();
    
    // Пошук ключових слів для загрози балістики
    if (lowerMessage.contains('загроза балістики') ||
        lowerMessage.contains('загроза балистики') ||
        lowerMessage.contains('балістична загроза') ||
        lowerMessage.contains('ballistic threat') ||
        lowerMessage.contains('пуск') && lowerMessage.contains('балістик')) {
      return BallisticAlertType.threat;
    }
    
    // Пошук ключових слів для відбою
    if (lowerMessage.contains('відбій загрози балістики') ||
        lowerMessage.contains('відбій балістики') ||
        lowerMessage.contains('відбій балістичної') ||
        lowerMessage.contains('ballistic all clear')) {
      return BallisticAlertType.allClear;
    }
    
    return null;
  }
  
  /// Витягти назву регіону з повідомлення
  static String? extractRegion(String message) {
    // Типові паттерни: "Загроза балістики! Київська область"
    // або "Київська область - загроза балістики"
    
    final regionPatterns = [
      RegExp(r'([\w\-]+ська область)', caseSensitive: false),
      RegExp(r'(м\.\s*Київ)', caseSensitive: false),
      RegExp(r'(Київ(?:ська)?)', caseSensitive: false),
    ];
    
    for (final pattern in regionPatterns) {
      final match = pattern.firstMatch(message);
      if (match != null) {
        return match.group(1);
      }
    }
    
    return null;
  }
}

enum BallisticAlertType {
  threat,
  allClear,
}

File: neptun_expo_app/src/features/map/services/ballisticAlertService.ts
/** Mirrors Flutter `BallisticAlertService`. */
export type BallisticAlertType = 'threat' | 'allClear';

const MIN_THREAT_DURATION_MS = 10_000;

type RegionCallback = (region: string | null) => void;

class BallisticAlertServiceImpl {
  private threatCallbacks: RegionCallback[] = [];
  private allClearCallbacks: RegionCallback[] = [];
  private _active = false;
  private _region: string | null = null;
  private _lastThreatAt: number | null = null;

  get isActive(): boolean {
    return this._active;
  }

  get currentRegion(): string | null {
    return this._region;
  }

  onThreat(cb: RegionCallback): () => void {
    this.threatCallbacks.push(cb);
    return () => {
      this.threatCallbacks = this.threatCallbacks.filter((x) => x !== cb);
    };
  }

  onAllClear(cb: RegionCallback): () => void {
    this.allClearCallbacks.push(cb);
    return () => {
      this.allClearCallbacks = this.allClearCallbacks.filter((x) => x !== cb);
    };
  }

  triggerThreat(region?: string | null): void {
    this._active = true;
    this._region = region ?? null;
    this._lastThreatAt = Date.now();
    for (const cb of this.threatCallbacks) cb(this._region);
  }

  triggerAllClear(region?: string | null): void {
    if (this._lastThreatAt != null && Date.now() - this._lastThreatAt < MIN_THREAT_DURATION_MS) {
      return;
    }
    this._active = false;
    this._region = null;
    for (const cb of this.allClearCallbacks) cb(region ?? null);
  }

  dismissThreat(): void {
    this._active = false;
    this._region = null;
    this._lastThreatAt = null;
  }

  static detectAlertType(message: string): BallisticAlertType | null {
    const lower = message.toLowerCase();
    if (
      lower.includes('загроза балістики') ||
      lower.includes('загроза балистики') ||
      lower.includes('балістична загроза') ||
      lower.includes('ballistic threat') ||
      (lower.includes('пуск') && lower.includes('балістик'))
    ) {
      return 'threat';
    }
    if (
      lower.includes('відбій загрози балістики') ||
      lower.includes('відбій балістики') ||
      lower.includes('відбій балістичної') ||
      lower.includes('ballistic all clear')
    ) {
      return 'allClear';
    }
    return null;
  }

  static extractRegion(message: string): string | null {
    const patterns = [
      /([\w\-]+ська область)/i,
      /(м\.\s*Київ)/i,
      /(Київ(?:ська)?)/i,
    ];
    for (const re of patterns) {
      const m = message.match(re);
      if (m?.[1]) return m[1];
    }
    return null;
  }
}

export const ballisticAlertService = new BallisticAlertServiceImpl();
